include/EntryPointConfirmOptIn.php

```diff
@@ -43,40 +43,104 @@
 /**
  * Class confirm_opt_in
  */
-class ConfirmOptIn
+class EntryPointConfirmOptIn
 {
     /**
      * @var EmailAddress $emailAddress
      */
     private $emailAddress;
+    
+    /**
+     * 
+     * @param array $request
+     * @param array $post
+     */
+    public function __construct($request = null, $post = null) {
+        
+        if(is_null($request)) {
+            $request = $_REQUEST;
+        }
+        
+        if(is_null($post)) {
+            $post = $_POST;
+        }
+        
+        $method = isset($request['method']) && $request['method'] ? $request['method'] : null;
```

Comment:
If you someone create a custom entry point, then add a custom name for it. I don't understand the problem. Please explain.

Reply:
My concern is related the inversion of control, do not repeat and that it is uncommon to use a constructor in this fashion.

A third party developer is likely to to want just to call methods from the entrypoint; because it is the usual pattern we use in SuiteCRM. Since we are using classes, they would expect the following pattern:

```php
<?php
$foo = new foo();  // instantiate class
$foo->doBar(): // Do something with class
if ($foo->isSomething()) {
    $foo->doBuzz(): // Do something else with class
}
```

conflating the constructor with the overall process is a confusing pattern, because we would expect the construction to be separate from calling the methods in the "controller" to control the flow.

Just to be clear, there is nothing wrong with calling a method in the constructor. However, you would expect the construct to set up the class, then call something in the class, and then deal with the response. The flow of control should be in the entrypoint file. So that we do not need to repeat the code in the customisation. 

I also have no problem with using a controller. provided that we follow the pattern. This is why I am suggesting to have a separate class for the view. and move any non PHP code into the template.

So can we move some of this code in to other methods eg. execute/process? Can we keep the view and business logic separate?

This is also in relation to the sugar_cleanup call. Please see my comment below regarding the sugar_cleanup. Since github comments are related to the line of code.

Reply:
I believe that we using OOP. Your example isn't but the next one is:
```
Buzzer.php
<?php
class Buzzer {
    public function __construct() {
        $foo = new foo();  // instantiate class
        $foo->doBar(): // Do something with class
        if ($foo->isSomething()) {
            $foo->doBuzz(): // Do something else with class
        }
    }
}

BuzzerCall.php:
<?php
new Buzzer();
```

now you still can make a CustomBuzzer, for e.g:
```
<?php
new CustomBuzzer { ... }
```
or you can implement any custom php as in your example above. Nothing to changed.

Reply:
My code example represents entrypoint file.

Controlling the process in a entrypoint file is the pattern used in SuiteCRM. If we are building a controller, we should follow the same pattern. Take the usage SugarApplication::execute as an example.

In the execute method, it set up a controller and executes it's action. This is what I mean by confusing; because in the current convention, the construct and the caller, and the action are separated. However, in your code the are all combined.

Furthermore, the https://github.com/salesagility/SuiteCRM/pull/4932/files#diff-25e9e5d4bad721b8ee020c27be3e7851L278 - could add a check to the switch case so you that it calls the the method dynamically? that would solve this issue.

Reply:
outdated. see previous comment and sourcecode